File: src/main/old/bbarena/model/event/game/NewHalfEvent.java
package bbarena.model.event.game;

import bbarena.model.Arena;
import bbarena.model.TurnMarker;
import bbarena.model.util.Concat;

import java.util.ArrayList;
import java.util.List;

public class NewHalfEvent extends GameEvent {

    private static final long serialVersionUID = -1919966604189789283L;
    private int _half;
    private List<TurnMarker> _oldMarkers = new ArrayList<>();

    public NewHalfEvent() {
        super();
    }

    @Override
    public void doEvent(Arena arena) {
        _half = arena.getHalf();
        arena.setHalf(_half + 1);
        List<TurnMarker> turnMarkers = arena.getTurnMarkers();
        for (TurnMarker turnMarker : turnMarkers) {
            TurnMarker oldMarker = new TurnMarker(turnMarker.getTeam());
            oldMarker.setRerolls(turnMarker.getRerolls());
            oldMarker.setUsedReroll(turnMarker.isUsedReroll());
            oldMarker.setTurn(turnMarker.getTurn());
            oldMarker.setStatus(turnMarker.getStatus());
            _oldMarkers.add(oldMarker);

            turnMarker.setTurn(0);
            turnMarker.setUsedReroll(false);
            turnMarker.setRerolls(turnMarker.getTeam().getReRolls());
        }
    }

    @Override
    protected void undoEvent(Arena arena) {
        arena.setHalf(_half);

        for (int i = 0; i < arena.getTurnMarkers().size(); i++) {
            TurnMarker oldMarker = _oldMarkers.get(i);
            TurnMarker turnMarker = arena.getTurnMarkers().get(i);
            turnMarker.setStatus(oldMarker.getStatus());
            turnMarker.setTurn(oldMarker.getTurn());
            turnMarker.setUsedReroll(oldMarker.isUsedReroll());
            turnMarker.setRerolls(oldMarker.getRerolls());
        }
    }

    @Override
    public String getString() {
        return Concat.INSTANCE.buildLog(getClass());
    }

}

File: src/main/kotlin/bbarena/model/event/game/NewHalfEvent.kt
package bbarena.model.event.game

import bbarena.model.Arena
import bbarena.model.TurnMarker

class NewHalfEvent() : GameEvent() {

    private val oldMarkers = mutableListOf<TurnMarker>()

    override fun doEvent(arena: Arena) {
        arena.half += 1

        // TODO: move turnmarker reset to a different event
        val turnMarkers = arena.turnMarkers
        for (turnMarker in turnMarkers) {
            val oldMarker = TurnMarker(turnMarker.team)
            oldMarker.rerolls = turnMarker.rerolls
            oldMarker.isUsedReroll = turnMarker.isUsedReroll
            oldMarker.turn = turnMarker.turn
            oldMarker.status = turnMarker.status
            oldMarkers.add(oldMarker)

            turnMarker.turn = 0
            turnMarker.isUsedReroll = false
            turnMarker.rerolls = turnMarker.team.reRolls
        }
    }

    override fun undoEvent(arena: Arena) {
        arena.half -= 1

        for (i in 0 until arena.turnMarkers.size) {
            val oldMarker = oldMarkers[i]
            val turnMarker = arena.turnMarkers[i]
            turnMarker.status = oldMarker.status
            turnMarker.turn = oldMarker.turn
            turnMarker.isUsedReroll = oldMarker.isUsedReroll
            turnMarker.rerolls = oldMarker.rerolls
        }
    }

}
